torch/csrc/dynamo/cache_entry.h

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+#include <torch/csrc/utils/python_compat.h>
+
+#ifdef __cplusplus
+
+#include <torch/csrc/utils/pybind.h>
+#include <list>
+
+namespace py = pybind11;
+
+extern "C" {
+
+#endif
+
+/*
+Our cache resides on the extra scratch space of the code object. The structure
+of the cache is as follows:
+
+-> ExtraState
+  -> CacheEntry (list)
+    -> check_fn
+    -> code
+  -> FrameState
+
+CacheEntry is a linked list node containing the check_fn for guards
+and the optimized code.
+
+The FrameState is a PyDict that enables sharing between different frames. This
+is used to detect dynamism in automatic dynamic shapes.
+
+These two are encapsulated into a ExtraState.
+*/
+
+typedef struct CacheEntry CacheEntry;
+typedef struct ExtraState ExtraState;
+
+#ifdef __cplusplus
+
+typedef struct CacheEntry {
+  // check the guards: lambda: <locals of user function>: bool
+  py::object check_fn;
```

Comment:
Have you benchmarked this?  This (and code) are the only part on the critical path, so want to much sure making it a py::object doesn't slow things down.

You can use ./benchmarks/dynamo/microbenchmarks/overheads.py

Reply:
My results are:
```
pre changes
eager    6.4us (warmup=0.0s)
compiled 15.0us (warmup=8.3s)

post changes
eager    6.2us (warmup=0.0s)
compiled 14.9us (warmup=9.1s)
```

Reply:
Yes seems fine, so just need to address test failures.